samcli/lib/package/utils.py

```diff
@@ -176,8 +175,15 @@ def upload_local_artifacts(
 
     local_path = make_abs_path(parent_dir, local_path)
 
-    # Or, pointing to a folder. Zip the folder and upload (zip_method is changed based on resource type)
     if is_local_folder(local_path):
+        files = os.listdir(local_path)
+        # Custom build methods may automatically produce a
+        # Zip.
+        if len(files) == 1:
+            archive_path = os.path.join(local_path, files[0])
+            if is_zip_file(archive_path):
+                return uploader.upload_with_dedup(archive_path)
+        # Or, pointing to a folder. Zip the folder and upload (zip_method is changed based on resource type)
```

Comment:
Thanks for your contribution! Can we add some logging here to track this down later?